### PR TITLE
[WIP] Eurus

### PR DESCRIPTION
**Pull request by AlexPiche**
## Tests
- [x] Reproduce large DeepSeek run

## Changes
- [x] Handle tensor and pipeline parallelism more gracefully.
- [x] Easily configurable task template and system prompt. 
- [x] Train on the [PRIME RL dataset. ](https://huggingface.co/datasets/PRIME-RL/Eurus-2-RL-Data)
- [x] Remove `base_url: str | list[str]` for `base_url: str` in the LLM class.
- [x] Comment out `symbolic_equal` since it was hanging and it is not compatible with our current version of hydra.
- [x] Improved and clean stats. 
- [x] Implemented a `is_debug_mode()` in `VLLMServiceManager`
- [x] Add multiprocess support for dataset.map in `finetune/data.py`
- [x] Rename `datasets` to `datasets_list` to avoid conflict with the `datasets` library.
- [x] Normalize the loss to be over sequences and not tokens, i.e. replace `masked_mean` by `masked_sum`.
- [x] Refactor `load_datasets`. 
- [x] Do not count the token of the prompt at every call. 
- [x] Use the vllm tokenization for `llm_call.prompt_length_tokens` and `llm_call.output_length_tokens`
- [x] Assert that LLM does not produce more than `cfg.llm.parameters.max_tokens`
- [x] Truncate from the left long prompts

### CODE REVIEW

File: conf/rl_eurus.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+llm:
+  parameters:
+    # CoT are much longer, but the model only has 4096 tokens context
+    max_tokens: 3072
+
+test_every_n_iterations: 1
+vllm_config:
+  vllm_kwargs:
+    --gpu-memory-utilization: 0.99
+    --max-num-seqs: 1024
```
Comment (rizar) [inline conf/rl_eurus.yaml:19]: Might be better to use a smaller number, cause surely we can't have enough KV cache for 4096 x 1024 tokens on 1 GPU.

File: examples/rl_gsm8k/cot_math_agent.py [hunk 2]
```diff
@@ -75,7 +80,7 @@ def create(cls, llm: LLM):
                 ReasoningNode(
                     name="cot",
                     agent_step_cls=MathAgentStep,
-                    store_llm_calls=True,
+                    system_prompt=EURUS_SYSTEM_PROMPT,
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:83]: Can we make this configurable?

File: examples/rl_gsm8k/orchestrate_rl.py [hunk 3]
```diff
@@ -143,15 +146,26 @@ def extract_tape_training_samples(
...
+        llm_call = step.metadata.other["llm_call"]
+
+        if isinstance(llm_call, dict):
+            llm_call = LLMCall(**llm_call)
+
+        tape_prompt_tokens += llm_call.prompt_length_tokens
+        tape_output_tokens += llm_call.output_length_tokens
+
+        if split_name == "train":
+            if llm_call.output_length_tokens >= cfg.llm.parameters.max_tokens:
```
Comment (rizar) [inline examples/rl_gsm8k/orchestrate_rl.py:165]: how can this happen, how can LLM return more than the max number of tokens?

File: examples/rl_gsm8k/utils.py [hunk 4]
```diff
@@ -60,7 +62,10 @@ def __init__(
         self.port = port
         self.ports = []
         self.processes = []
-        self.gpus_per_model_instance = gpus_per_model_instance
+        pipeline_parallel_size = kwargs.get("--pipeline-parallel-size", 1)
```
Comment (rizar) [inline examples/rl_gsm8k/utils.py:65]: are you using pipeline parallelism or tensor parallelism in your experiments now?

File: examples/rl_gsm8k/utils.py [hunk 5]
```diff
@@ -394,11 +396,15 @@ def launch_training(
 
     logger.info(f"Launching training with command: {' '.join(base_cmd)}")
     try:
+        # set env variable TORCH_NCCL_ENABLE_MONITORING=0
+        env = os.environ.copy()
+        env["TORCH_NCCL_ENABLE_MONITORING"] = "0"
```
Comment (rizar) [inline examples/rl_gsm8k/utils.py:401]: what is this for?

File: tapeagents/finetune/data.py [hunk 6]
```diff
@@ -216,16 +218,24 @@ def create_dataloader(
 
         logger.info(f"Raw data part size: {dataset_part.num_rows}")
         logger.info(f"Raw data part fingerprint: {dataset_part._fingerprint}")
-        dataset_part = dataset_part.map(preprocess, keep_in_memory=True, load_from_cache_file=False)
+
+        num_proc = (os.cpu_count() // accelerator.num_processes) or 1
```
Comment (rizar) [inline tapeagents/finetune/data.py:222]: I suppose this code can break some prompt groups, can it?

**Review by rizar — COMMENTED** [review]
nice progress! some comments...

File: tapeagents/llms.py [hunk 7]
```diff
@@ -667,6 +666,8 @@ def batch_generate(self, prompts: list[Prompt]) -> list[LLMCall]:
             self.tokenizer.apply_chat_template(p.messages, add_special_tokens=True, add_generation_prompt=True)
             for p in prompts
         ]
+        if self.max_prompt_length > 0:
```
Comment (rizar) [inline tapeagents/llms.py:669]: I don't think we want this in TapeAgents, it's better to have prompt trimming in the downstream example code.

File: conf/rl_eurus.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+defaults:
+  - rl_gsm8k
+  - _self_
+model_path: /mnt/llmd/base_models/Eurus-2-7B-SFT
+dataset_name: eurus
+output_dir: outputs/rl_eurus
+finetune:
+  max_seq_len: 4096
```
Comment (rizar) [inline conf/rl_eurus.yaml:8]: Are you sure `finetune` actually consumes this parameter?

File: examples/rl_gsm8k/utils.py (hunk 4 above)
Comment (AlexPiche) [reply examples/rl_gsm8k/utils.py:62]: I did not end up using parallelism, but I updated the code to handle both pipeline and tensor parallelism more gracefully.

File: examples/rl_gsm8k/orchestrate_rl.py (hunk 3 above)
Comment (AlexPiche) [reply examples/rl_gsm8k/orchestrate_rl.py:172]: `llm_call.output_length_tokens` is the number of tokens produced by vLLM and the special tokens of the chat template, e.g., `<|im_end|>\n`.

File: examples/rl_gsm8k/cot_math_agent.py [hunk 9]
```diff
@@ -62,18 +67,53 @@ def parse_completion(self, completion: str, prompt_id: str) -> Generator[Step, N
...
+            1. Cleans the tape content
+            2. Gets steps description
+            3. Converts tape to messages
+            4. Checks token count and trims if needed
+            5. Reconstructs messages if trimming occurred
+        """
+        cleaned_tape = self.prepare_tape(tape)
+        steps_description = self.get_steps_description(tape, agent)
+        messages = self.tape_to_messages(cleaned_tape, steps_description)
+        messages = self.tape_to_messages(cleaned_tape, steps_description)
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:96]: We don't need to copy-paste all this code. Let's simplify.

Comment (rizar) [reply examples/rl_gsm8k/cot_math_agent.py:96]: I imagine that all you need in this particular case is just to make the one and only user message from the one and only step 0 of the tape.

File: examples/rl_gsm8k/cot_math_agent.py [hunk 10]
```diff
@@ -62,18 +67,53 @@ def parse_completion(self, completion: str, prompt_id: str) -> Generator[Step, N
...
+        agent.llm.load_tokenizer()
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:97]: it's cleaner to do this before just when you create the LLM

File: conf/rl_eurus.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+model_path: /mnt/llmd/base_models/Eurus-2-7B-SFT
+dataset_name: eurus
+output_dir: outputs/rl_eurus
+attempts: 1
+llm:
+  parameters:
+    # CoT are much longer, but the model only has 4096 tokens context
+    max_tokens: 3072
+
+# EURUS already apply this template: {task}\n\nPresent the answer in LaTex format: \\boxed{Your answer}
```
Comment (rizar) [inline conf/rl_eurus.yaml:13]: What is the meaning of this comment here?

File: examples/rl_gsm8k/cot_math_agent.py [hunk 12]
```diff
@@ -62,21 +67,35 @@ def parse_completion(self, completion: str, prompt_id: str) -> Generator[Step, N
             return
         yield step
 
+    def make_prompt(self, agent: Any, tape: Tape) -> Prompt:
+        messages = []
+        if self.system_prompt:
+            messages.append({"role": "system", "content": self.system_prompt})
+        messages.append({"role": "user", "content": tape.steps[0].llm_view()})
+        #messages = self.tape_to_messages(cleaned_tape, steps_description)
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:79]: is the comment still needed?

File: examples/rl_gsm8k/cot_math_agent.py [hunk 13]
```diff
@@ -62,21 +67,35 @@ def parse_completion(self, completion: str, prompt_id: str) -> Generator[Step, N
...
-    def create(cls, llm: LLM):
+    def create(cls, system_prompt: str, llm: LLM, max_prompt_length: int):
         agent = super().create(
             llm,
             nodes=[
                 ReasoningNode(
                     name="cot",
                     agent_step_cls=MathAgentStep,
-                    store_llm_calls=True,
+                    system_prompt=system_prompt if system_prompt else "",
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:97]: `system_prompt` is mandatory argument in this function, what is the purpose of this check?

File: examples/rl_gsm8k/deepseek_math_eval/eval_utils.py [hunk 14]
```diff
@@ -299,8 +300,9 @@ def math_equal(
         if call_with_timeout(symbolic_equal_process, prediction, reference):
             return True
     else:
-        if symbolic_equal(prediction, reference):
-            return True
+        pass
```
Comment (rizar) [inline examples/rl_gsm8k/deepseek_math_eval/eval_utils.py:303]: it would be nice to add the reason as to why we've comment this out, which as far as I remember is that this code hangs

File: tapeagents/llms.py [hunk 15]
```diff
@@ -727,7 +733,20 @@ def batch_generate(self, prompts: list[Prompt]) -> list[LLMCall]:
                 logger.exception(f"Failed to parse llm response: {r}")
                 raise e
             output = LLMOutput(content=content)
-            llm_call = self.log_output(prompts[i], output)
+            # if logprobs is not None, we will directly take the token counts from vLLM
+            # otherwise, we will count the tokens in the output using the tokenizer (which is sometimes inaccurate)
+            if logprobs:
+                llm_call = self.log_output(prompts[i], output, count_tokens=False)
+                llm_call.prompt_length_tokens = len(prompt_token_ids[i])
```
Comment (rizar) [inline tapeagents/llms.py:740]: nice one!

**Review by rizar — APPROVED** [review]
Start looking pretty good! you can merge once you address all comments.
